Dev/forms and requests

krzysztofhewelt opened the pull request.

krzysztofhewelt committed:
Add react-hook-form, yup and react query. Fix logo and styles

krzysztofhewelt committed:
Implement form validation, build pages, improve layout responsivity

krzysztofhewelt commented on ui/src/components/Drawer.tsx:
```diff
@@ -1,48 +1,63 @@
+          <DrawerLink path={paths.TASKSDONE} textKey="drawer.tasksFinished" onClick={closeDrawer} />
+          <DrawerLink path={paths.TASKSTODO} textKey="drawer.checkTasks" onClick={closeDrawer} />
+          <DrawerLink path={paths.PLAYGROUND} textKey="drawer.sandbox" onClick={closeDrawer} />
+          <a href="" onClick={logout}>
```
Nie mogę tego zrobić jako DrawerLink bo musiałbym zrobić komponent, opłaca się?

amelcer replied:
jeśli style są takie same to może zostać

krzysztofhewelt commented on ui/src/components/Header.tsx:
```diff
@@ -0,0 +1,20 @@
+    <div className="mb-4 flex flex-col leading-none">
+      {t('header.welcome')}
+      <span className="block text-xl font-bold">{username}</span>
+      {t('header.mainPage')} {location.key}
```
Key za każdym rebuildem się zmienia. Jedynie co można zrobić to pathname albo state dla każdego route.

amelcer replied:
no to możemy użyć pathname. Mamy już zdefiniowany obiekt z wszystkimi ścieżkami to możemy na jego podstawie stworzyc kolejny obiekt który będzie wskazywał na klucze translacji. Coś typu 

```ts
import paths from '@routes/paths';

export const pageNames = {
 [paths.PLAYGROUND]: 'pageNames.palyground',
  [paths.TASKSTODO]: 'pageNames.taskTodo',
} as const
```

i potem użyć go tutaj 

```suggestion
      {t(pageNames[location.path])}
```

krzysztofhewelt commented on ui/src/components/Input.tsx:
```diff
@@ -1,42 +1,37 @@
       {children}
-      <div className="text-red-500">{error}</div>
+      <div className="text-left text-red-500">
+        {errors?.[name] && <div>{t(errors[name]?.message?.toString() || 'validation.error')}</div>}
```
Nie jestem przekonany co do tego zapisu (validation.error - musimy alternatywnie zapewnić jakiś string bo inaczej krzyczy, że może być undefined), lecz nie powoduje warningów. Jak inaczej to zrobić?

amelcer replied:
tutaj nie powinniśmy wywoływać funkcji t z atrybutem message. Message jest już zwracany jako przetłumaczony string. 
```suggestion
        {errors?.[name] && <div>{errors[name]?.message?.toString() || t('validation.error')}</div>}
```

amelcer replied:
można przygotować dodatkową funckje która będzie sprawdzać jakiego typu jest `errors[name]?.message`, ale ja nie widzę problemu żeby to zostawic w ten sposób. To że krzyczy że może być undefined to właściwie dobrze, przynajmniej wymusza na nas żeby zapewnić jakąś domyślną wiadomośc

krzysztofhewelt commented on ui/src/components/PickerDifficultyLevel.tsx:
```diff
@@ -1,8 +1,8 @@
-import { difficultyLevels } from '../consts/difficultyLevel.ts';
-import { useTranslation } from 'react-i18next';
+import { difficultyLevels } from 'consts/difficultyLevel';
+import { t } from 'i18next';
```
To jak się korzysta z tego i18n: useTranslation czy import { t } ? Bo można to też tak zrobić...

amelcer replied:
lepiej korzystać z t zwracanego przez hook useTranslation. Wtedy mamy pewność że po zmianie języka jest odpowiednia funkcja wywołwyana

krzysztofhewelt commented on ui/src/libraries/downloadImage.tsx:
```diff
@@ -0,0 +1,6 @@
+export const downloadImage = (dataBase64: string, name = 'rysunek.png') => {
+  const downloadLink = document.createElement('a');
+  downloadLink.href = dataBase64;
```
Na razie bez konwersji do blob...

krzysztofhewelt commented on ui/src/pages/playground/playground.tsx:
```diff
@@ -0,0 +1,11 @@
+export default function Playground() {
+  return (
+    <>
+      <div className="mx-auto h-[750px] w-3/4">
```
W ogóle nie reaguje na procentowe wartości wysokości, trzeba na sztywno. Odporne na divy z height 100vh.

amelcer replied:
to dlatego że na praent element masz ustawione `min-height` a nie `height` przez to nie ma na podstawie czego obliczać procentów

krzysztofhewelt commented on ui/src/pages/tasks/taskdraw.tsx:
```diff
@@ -0,0 +1,33 @@
+
+      {id}
+
+      <div className="text-right text-2xl font-bold">TIMER</div>
```
Timer do zrobienia

krzysztofhewelt commented on ui/src/pages/tasks/tasklist.tsx:
```diff
@@ -0,0 +1,78 @@
+            description={el.description}
+            image={el.image}
+            key={el?.taskId}
+            onClick={() => navigate(paths.TASKDRAW, { id: el.taskId })}
```
Nie można tego tak zrobić, daje warningi. Jedyne rozwiązanie, które przedstawiają to odpowiednik: nagivate(`/tasks/${el.taskId}`)

amelcer replied:
to dlatego że ta funckja oczekuje innych parametrów. 
![image](https://github.com/krzysztofhewelt/drawventure/assets/60044260/dadbb1e3-4ece-436d-8ee8-334330903e93)

Jeśli chcemy to zrobić poprawnie to trzeba użyć `generatePath` z react-router-dom i potem

```suggestion
              onClick={() => navigate(generatePath(paths.TASKDRAW, {id: el.taskId}))}
```

krzysztofhewelt commented on ui/src/yupi18n.ts:
```diff
@@ -0,0 +1,11 @@
+import { setLocale } from 'yup';
```
Moim zdaniem średnio przydatne, bo nie zapewnia obsługi swoich pól: np. przy regex'ie wyświetla całego regexa, a też nie wypadałoby pisać, że "hasło jest niepoprawne" nie wiadomo dlaczego.

amelcer replied:
czemu średnio przydatne? Jeśli pole będzie wymagane to wolisz za każdym razem podawać jakiego klucza ma użyć? Daje to właśnie możliwość że wprowadzamy domyślne wiadomości które używają translacji, jeśli potrzebujemy użyć innej wiadomości to nie ma problemu żeby ją zdefiniować podczas tworzenia schemy.

krzysztofhewelt commented on ui/src/pages/tasks/tasklist.tsx:
```diff
@@ -0,0 +1,78 @@
+import { useNavigate } from 'react-router-dom';
+import paths from '@routes/paths';
+
+export default function TaskList() {
```
Może być uniwersalne zarówno dla zadań do wykonania i wykonanych. Zrobić komponent, czy po state/pathname robić requesty inne?

amelcer replied:
Komponent TaskList powinien odpowiadać tylko za logikę wyświetlania zadań. W jakim kontekscie będzie użyty nie powinno go interesować. Tak więc powinien przyjmować tylko jakie zadania wyświetlić i potem w innych komponentach będziemy go odpowiednio używać z propami

krzysztofhewelt commented on ui/src/components/DifficultyLevel.tsx:
```diff
@@ -1,13 +1,14 @@
 import classNames from 'classnames';
+import { difficultyLevels } from 'consts/difficultyLevel';
 
 const DifficultyLevel = ({ difficulty }: { difficulty: number }) => {
```
Typ number czy DifficultyLevels? Wtedy jak z liczby podać jako argument z requestu?

amelcer replied:
trochę nie rozumiem pytania. Możesz rozjaśnić?

krzysztofhewelt commented on ui/src/components/Card.tsx:
```diff
@@ -7,20 +7,20 @@ interface Props {
   return (
     <div className={style} onClick={onClick}>
-      <img src={image} alt={image} className="my-auto h-3/4 w-1/3" />
+      <img src={image} alt={image} className="my-auto h-48" />
```
Musiałem dać wysokość na sztywno bo inaczej rysunek rozszerza diva. Jak zrobić flex o fixed height ale tak żeby rysunek nie powiększał diva.

krzysztofhewelt commented on ui/src/pages/home/home.tsx:
```diff
@@ -1,29 +1,25 @@
+      <div className="flex max-h-screen min-h-screen flex-col gap-10">
         <Card
-          className="h-40 rounded-normal bg-white shadow-extra hover:cursor-pointer"
+          className="flex-1 rounded-normal bg-white shadow-extra hover:cursor-pointer"
```
Divy powinny mieć po równo, dopóki rysunek nie powiększy diva, image powinien mieć maksymalny height diva, ale to nie działa - robi max height obrazka.

amelcer committed:
Resolve imports order

krzysztofhewelt committed:
Build alpha prototype

TODO:
- account activation and password reset pages
- change password page
- no tasks done information
- fix Header pageNames match with params route
- improve TaskCard description for showing results
- react-query and requests to Firebase
- loading view
- error pages (404, 500, ...)
- privacy policy page
- colorful timer depends on best time
- scoreboard to design and implementation
- animations

amelcer commented on ui/src/components/Form.tsx:
```diff
@@ -0,0 +1,32 @@
+  return (
+    <FormProvider {...formMethods}>
+      <form className={className} onSubmit={formMethods.handleSubmit(onSubmit)}>
+        {Array.isArray(children)
+          ? children.map((child) => {
+              return child.props.name
+                ? createElement(child.type, {
+                    ...{
+                      ...child.props,
+                      key: child.props.name,
+                    },
+                  })
+                : child;
+            })
+          : children}
```
```suggestion
       {children}
```

Samo children wystarczy. Nie ma potrzeby tak tego rozdzielać

amelcer commented on ui/src/components/TasksList.tsx:
```diff
@@ -0,0 +1,54 @@
+  const [tasksFilter, setTaskFilter] = useState(filterTasksByDifficulty(tasks, difficultyLevel));
+  const navigate = useNavigate();
+
+  useEffect(() => {
+    setTaskFilter(tasks.filter((task) => task.difficulty === difficultyLevel));
+  }, [difficultyLevel, tasks]);
```
to też niepotrzebne, możemy od razu w linii 38 robić task.filter().map() i od razu to renderować. Ten useEffect wprowadza tutaj niepotrzebny chaos

amelcer commented on ui/src/components/Timer.tsx:
```diff
@@ -0,0 +1,37 @@
+import { useEffect, useState } from 'react';
+
+const Timer = () => {
```
zamiast komponentu proponwałbym stworzenie hooka, potem będzie łatwiej to wysłać na backend w komponencie który będzie używał tego timera

amelcer commented on ui/src/components/Timer.tsx:
```diff
@@ -0,0 +1,37 @@
+import { useEffect, useState } from 'react';
+
+const Timer = () => {
+  const [minutes, setMinutes] = useState('00');
+  const [seconds, setSeconds] = useState('00');
```
po co trzymać minuty i sekundy osobno? Proponowałbym stworzyć obiekt daty który będzie naszym startem (lub chyba nawet lepiej przyjąć go jako argument) i potem obliczać różnice między nim a obecną datą. Następnie wyświetlać to przy pomocy obiektu [Intl](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/DateTimeFormat/format) jeśli nie chcemy używać żadnej bilbioteki do dat. Chociaż tutaj też proponowałbym użycie np https://date-fns.org/ łatwiej wtedy będzie obliczać róznice w czasie i wyświetlać w określonych formatach

amelcer commented on ui/src/consts/regex.ts:
```diff
@@ -0,0 +1,6 @@
+export const regex = {
+  email: new RegExp(
+    /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/
+  ),
+  password: new RegExp(/^(?=.*[A-Z])(?=.*[a-z])(?=.*[0-9])(?=.*[!@#$%^&*])[A-Za-z0-9!@#$%^&*]{8,128}$/),
```
to moglibyśmy stworzyć przy pomocy yupa.

yup.string().min(8, 'hasło musi mieć przynajmniej 8 znaków').max(200).matches(/tutaj regex tylko na sprawdzenie czy ma znak specjalny)

amelcer commented on ui/src/consts/regex.ts:
```diff
@@ -0,0 +1,6 @@
+export const regex = {
+  email: new RegExp(
```
yup.string().mail()  jest wystarczającym walidatorem, nie ma potrzeby bawić się w swoje skomplikowane regexy. Email po bez kropki jest jaknajbardziej prawidłowy
https://en.wikipedia.org/wiki/Email_address#Examples
`admin@example` dla lokalnych środowisk, lub nawet czysto teoretycznie ktoś może mieć email w top level domenie np admin@com

krzysztofhewelt committed:
Prepare useTimer hook

amelcer committed:
Rename paths

amelcer committed:
Handle dynamic page names

krzysztofhewelt committed:
Fix eraser mode

amelcer committed:
Move use timer to hooks
